validator/src/journal/block_manager.rs

```diff
@@ -0,0 +1,1039 @@
+            .map(|anchor| anchor.blockstore_name.as_str())
+    }
+
+    fn convert(&mut self, blocks: Vec<Block>) -> Vec<(Block, i64, i64)> {
+        blocks
+            .into_iter()
+            .map(|block| {
+                let anchor = self.anchors_by_block_id
+                    .remove(block.header_signature.as_str())
+                    .unwrap();
+                let anchors = self.anchors_by_blockstore_name
+                    .get_mut(anchor.blockstore_name.as_str())
+                    .expect("Anchors in anchor_by_block_id and anchors_by_blockstore_name lost integrity");
+                anchors.remove(block.header_signature.as_str());
+                (block, anchor.internal_ref_count, anchor.external_ref_count)
+            })
+            .collect()
+    }
+
+    fn iter_by_blockstore<'a>(
+        &'a self,
+        blockstore_name: &str,
+    ) -> Box<Iterator<Item = &Anchor> + 'a> {
+        let anchors_by_block_id = &self.anchors_by_block_id;
+
+        match self.anchors_by_blockstore_name.get(blockstore_name) {
+            Some(anchors) => {
+                let iter = anchors
+                    .iter()
+                    .map(move |block_id| anchors_by_block_id.get(block_id).unwrap());
+                Box::new(iter)
+            }
+            None => Box::new(::std::iter::empty()),
+        }
+    }
+
+    fn contains(&self, block_id: &str) -> bool {
+        self.anchors_by_block_id.contains_key(block_id)
+    }
+
+    fn increase_internal_ref_count(&mut self, block_id: &str) {
+        if let Some(anchor) = self.anchors_by_block_id.get_mut(block_id) {
+            anchor.internal_ref_count += 1;
+        }
+    }
+
+    fn decrease_internal_ref_count(&mut self, block_id: &str) {
+        if let Some(anchor) = self.anchors_by_block_id.get_mut(block_id) {
+            anchor.internal_ref_count -= 1;
+        }
+    }
+
+    fn get_internal_ref_count(&self, block_id: &str) -> Option<i64> {
+        self.anchors_by_block_id
+            .get(block_id)
+            .map(|anchor| anchor.internal_ref_count)
+    }
+
+    fn increase_external_ref_count(&mut self, block_id: &str) {
+        if let Some(anchor) = self.anchors_by_block_id.get_mut(block_id) {
+            anchor.external_ref_count += 1;
+        }
+    }
+
+    fn decrease_external_ref_count(&mut self, block_id: &str) {
+        if let Some(anchor) = self.anchors_by_block_id.get_mut(block_id) {
+            anchor.external_ref_count -= 1;
+        }
+    }
+
+    fn get_external_ref_count(&self, block_id: &str) -> Option<i64> {
+        self.anchors_by_block_id
+            .get(block_id)
+            .map(|anchor| anchor.external_ref_count)
+    }
+}
+
+struct Anchor {
+    pub blockstore_name: String,
+    pub external_ref_count: i64,
+    pub internal_ref_count: i64,
+    pub block_num: u64,
+    pub block_id: String,
+}
+
+#[derive(Default)]
+pub struct BlockManager {
+    block_by_block_id: HashMap<String, (Block, i64, i64)>,
```

Comment:
I would use a struct here instead of a tuple so it is clearer which i64 is for internal/external.

Reply:
Can we use a u64 here with checked subtraction to catch when we try to unref below 0? https://doc.rust-lang.org/std/primitive.u64.html#method.checked_sub